### ISSUE TITLE
Prepare for DNF 5

### ISSUE THREAD

**Issue by evan-goode** [calamares/calamares#2145]
DNF 5 is a new package manager that will replace DNF 4 in Fedora 39+. Starting in Fedora 39, the `dnf` command will be provided by the `dnf5` package rather than the `dnf` package, and `dnf5` will obsolete `dnf`. Since Calamares currently depends on DNF 4, it should choose one of the following strategies to avoid breaking the approaching Fedora upgrade and maintain compatibility with future versions of Fedora:
- Add support for DNF 5, and depend on the `dnf5` package in Fedora 39+ instead of dnf. The command-line interface isn't changing a whole lot between DNF 4 and DNF 5, so in Calamares's case, this change may be pretty quick. Builds of DNF 5 are available in this COPR repository: https://copr.fedorainfracloud.org/coprs/rpmsoftwaremanagement/dnf5-unstable, and documentation is available here: https://dnf5.readthedocs.io/en/latest/. You can call DNF 5 explicitly at `/usr/bin/dnf5`.
- Change the Calamares package to depend on `python3-dnf` instead of `dnf`, and call the `dnf-3` binary instead of `dnf`. The old DNF 4 command will still be available in the distribution, but only as `dnf-3`. (The binary is called `dnf-3`  rather than `dnf4` for historical reasons; it was originally the Python 3 version of DNF). The first option is preferred to this one; it is not recommended to modify installed software using both DNF 4 and DNF 5 on the same system.
  - I drafted up a small PR that should implement this change: https://github.com/calamares/calamares/pull/2144

We are planning to replace DNF with DNF5 in Fedora Rawhide very soon, in the next couple weeks, and the `calamares` package will break as long as it still depends on the `dnf` package.

After the necessary changes are made here in upstream, I'm happy to file the bugs in the Fedora issue tracker to get the package dependencies changed.

For more information about the switch to DNF 5, see https://fedoraproject.org/wiki/Changes/ReplaceDnfWithDnf5.

**Comment by kkofler** [comment]
Does DNF 5 need different CLI arguments than the ones we are currently passing? I was under the impression that is should be mostly CLI-compatible?

**Comment by kkofler** [comment]
If DNF 5 still insists on syncing the repositories even for a remove command, we will need `--cacheonly` before we can use it. Removing a local package should not require Internet access.

**Comment by kkofler** [comment]
I have merged your pull request.

**Comment by evan-goode** [comment]
Hi, as you may be aware, the switch to DNF 5 has been postponed, likely to Fedora 41. In Fedora 39, the `dnf` command will be provided by the `dnf` package, and DNF 5 will not be installed by default.

Calling `dnf-3` instead of `dnf` is still fine---you do not need to revert https://github.com/calamares/calamares/pull/2144.

In the meantime, we have set up a testing [COPR repository](https://copr.fedorainfracloud.org/coprs/rpmsoftwaremanagement/dnf5-testing/) that provides a version of DNF 5 that obsoletes DNF 4. It can be used to test software in an environment similar to the future release of Fedora (whichever that will be) when DNF 5 replaces DNF 4:

```
sudo dnf-3 copr enable rpmsoftwaremanagement/dnf5-testing
```

Enabling the COPR and upgrading your system should replace DNF 4 (the `dnf` package) with DNF 5 (`dnf5`), and `/usr/bin/dnf` will be DNF 5.

**Comment by evan-goode** [comment]
Just a heads-up, the Fedora 41 development cycle is beginning, and we are planning to obsolete DNF 4 by DNF 5 in Fedora Rawhide within the next few weeks. Again, libdnf is not going away, and the old DNF 4 command will still be available as `/usr/bin/dnf-3`, but `/usr/bin/dnf` will be DNF 5, which has a slightly different command-line interface. Packages should no longer depend on the `dnf` package in Fedora 41+.
